java/ql/lib/semmle/code/java/Constants.qll

```diff
@@ -0,0 +1,168 @@
+ * apply to compile-time constants.
+ *
+ * The input and output predicates are expected to be mutually recursive.
+ */
+module CalculateConstants<getBoolValSig/1 getBoolVal, getIntValSig/1 getIntVal> {
+  /** Gets the value of a constant boolean expression. */
+  pragma[assume_small_delta]
+  boolean calculateBooleanValue(Expr e) {
+    // No casts relevant to booleans.
+    // `!` is the only unary operator that evaluates to a boolean.
+    result = getBoolVal(e.(LogNotExpr).getExpr()).booleanNot()
+    or
+    // Handle binary expressions that have integer operands and a boolean result.
+    exists(BinaryExpr b, int left, int right |
+      b = e and
+      left = getIntVal(b.getLeftOperand()) and
+      right = getIntVal(b.getRightOperand())
+    |
+      (
+        b instanceof LTExpr and
+        if left < right then result = true else result = false
+      )
+      or
+      (
+        b instanceof LEExpr and
+        if left <= right then result = true else result = false
+      )
+      or
+      (
+        b instanceof GTExpr and
+        if left > right then result = true else result = false
+      )
+      or
+      (
+        b instanceof GEExpr and
+        if left >= right then result = true else result = false
+      )
+      or
+      (
+        b instanceof ValueOrReferenceEqualsExpr and
+        if left = right then result = true else result = false
+      )
+      or
+      (
+        b instanceof ValueOrReferenceNotEqualsExpr and
+        if left != right then result = true else result = false
+      )
+    )
+    or
+    // Handle binary expressions that have boolean operands and a boolean result.
+    exists(BinaryExpr b, boolean left, boolean right |
+      b = e and
+      left = getBoolVal(b.getLeftOperand()) and
+      right = getBoolVal(b.getRightOperand())
+    |
+      (
+        b instanceof ValueOrReferenceEqualsExpr and
+        if left = right then result = true else result = false
+      )
+      or
+      (
+        b instanceof ValueOrReferenceNotEqualsExpr and
+        if left != right then result = true else result = false
+      )
+      or
+      (b instanceof AndBitwiseExpr or b instanceof AndLogicalExpr) and
+      result = left.booleanAnd(right)
+      or
+      (b instanceof OrBitwiseExpr or b instanceof OrLogicalExpr) and
+      result = left.booleanOr(right)
+      or
+      b instanceof XorBitwiseExpr and result = left.booleanXor(right)
+    )
+    or
+    // Ternary expressions, where the `true` and `false` expressions are boolean constants.
+    exists(ConditionalExpr ce, boolean condition |
+      ce = e and
+      condition = getBoolVal(ce.getCondition()) and
+      result = getBoolVal(ce.getBranchExpr(condition))
+    )
+    or
+    // If a `Variable` is final, its value is its initializer, if it exists.
+    exists(Variable v | e = v.getAnAccess() and v.isFinal() |
+      result = getBoolVal(v.getInitializer())
+    )
```

Comment:
Should you consider adding SSA reads where you know the constant value of the SSA variable?

Reply:
Same for `calculateIntValue`.

Reply:
No, this file must include only exactly those steps that define compile-time-constants as per the Java Language Specification.

Reply:
I've tried to specify this in the qldoc for the module.

Reply:
I see that now, thanks.